### PR TITLE
Fixed multiple bugs on pipeline config save API - #1417, #1630

### PR DESCRIPTION
**Pull request by jyotisingh**

### CODE REVIEW

File: server/webapp/WEB-INF/rails.new/app/presenters/api_v1/config/tasks/on_cancel_representer.rb [hunk 1]
```diff
@@ -27,7 +27,7 @@ def from_hash(hash, options={})
 
         def to_hash(*options)
           return nil if on_cancel_config.getTask.getTaskType.eql?("killallchildprocess")
-          TaskRepresenter.new(on_cancel_config.getTask).to_hash(url_builder: UrlBuilder.new)
+          TaskRepresenter.new(on_cancel_config.getTask).to_hash(url_builder: nil)
```
Comment (zabil) [inline server/webapp/WEB-INF/rails.new/app/presenters/api_v1/config/tasks/on_cancel_representer.rb]: Why was this done?

File: config/config-api/src/com/thoughtworks/go/config/materials/PluggableSCMMaterialConfig.java [hunk 2]
```diff
@@ -315,11 +315,11 @@ public boolean equals(Object o) {
             return false;
         }
 
-        if (scmConfig == null || that.scmConfig == null) {
+        if (scmConfig != null ? !scmConfig.equals(that.scmConfig) : that.scmConfig != null) {
             return false;
         }
 
-        return this.getFingerprint().equals(that.getFingerprint());
```
Comment (zabil) [inline config/config-api/src/com/thoughtworks/go/config/materials/PluggableSCMMaterialConfig.java]: Ok, why are we removing this finger print check?

Comment (jyotisingh) [inline config/config-api/src/com/thoughtworks/go/config/materials/PluggableSCMMaterialConfig.java]: fingerprint is a computed field which internally checks for scmConfig itself. If you see, I have included scmConfig field for equals comparison and removed fingerprint, as is done for the other materials. Also, while saving a config xml, the scmConfig object could be null, and only the ref/scmId would be set. In such a case, equals comparison returned false for materials pointing to the same ref/scmId

File: server/webapp/WEB-INF/rails.new/app/presenters/api_v1/config/tasks/on_cancel_representer.rb (hunk 1 above)
Comment (jyotisingh) [inline server/webapp/WEB-INF/rails.new/app/presenters/api_v1/config/tasks/on_cancel_representer.rb]: `UrlBuilder` is  a test-only class. For production usage we need to pass along the controller. This was possibly added in error,  hence removing it. Also, a value of `nil` is sufficient as there are no embedded urls in TaskRepresenter or anything under it.

**Pull request merged**